### PR TITLE
Common function refactoring

### PR DESCRIPTION
**Pull request by JustinKuli**
Lots of refactors to make some of the common functions simpler. One goal was to increase readability of the tests, and to try to make more common behaviors more obvious in the future.

Some tests that enforce a policy would infrequently fail because the `Update` API call used would fail, because the object was modified before the action took place. The new common function that handles this will retry and shouldn't have those failures.

### CODE REVIEW

**Comment by JustinKuli** [comment]
/retest

Failure was (an unfortunately not uncommon flake):
```
------------------------------
• [FAILED] [306.421 seconds]
GRC: [P1][Sev1][policy-grc] Test the Policy Generator in an App subscription [BVT]
/go/src/github.com/stolostron/governance-policy-framework/test/integration/policy_generator_test.go:22
  [It] Sets up the application subscription
  /go/src/github.com/stolostron/governance-policy-framework/test/integration/policy_generator_test.go:41
  Begin Captured GinkgoWriter Output >>
    STEP: Verifying that the subscription-admin ClusterRoleBinding exists 08/18/22 16:36:08.646
    STEP: Cleaning up any existing subscription-admin user config 08/18/22 16:36:08.652
    STEP: Creating a subscription-admin user and configuring IDP 08/18/22 16:36:08.693
    STEP: Cleaning up the changes made to the cluster in the test 08/18/22 16:41:08.948
  << End Captured GinkgoWriter Output
  Timed out after 300.000s.
  Expected
      <*errors.errorString | 0xc0005bd3c0>: {
          s: "failed to login: Login failed (401 Unauthorized)\nVerify you have provided correct credentials.\n",
      }
  to be nil
  In [It] at: /go/src/github.com/stolostron/governance-policy-framework/test/integration/policy_generator_test.go:102
------------------------------
```

**Review by dhaiducek — APPROVED** [review]
/hold for reviews

Really good changes--thanks for updating all this! Along with this, we should probably open up an issue to update the README/docs here to explain the E2E test structure and about adding an E2E test.

File: test/common/policy_utils.go
Comment (JustinKuli) [reply test/common/policy_utils.go:66]: I can't find any in this repo that use Placement, but it's possible there are some in the policy-collection, which we test.

That brings up another point: in the policy-collection, the placement rules generally seem to be prepended with "placement-" as opposed to having "-plr" appended, like this function suggests. At some point we might want to align those, probably ensuring there's a consistent convention in the collection, and following that here.

**Comment by JustinKuli** [comment]
/retest 

```
  Timed out after 60.000s.
  Expected
      <v1.ComplianceState>: NonCompliant
  to equal
      <v1.ComplianceState>: Compliant
  In [It] at: /go/src/github.com/stolostron/governance-policy-framework/test/integration/policy_gatekeeper_operator_downstream_test.go:327
```

**Comment by JustinKuli** [comment]
/retest

Slightly suspicious... (second gatekeeper test fail, although it is a different one than before) ...
```
  Timed out after 180.001s.
  Expected
      <v1.ComplianceState>: NonCompliant
  to equal
      <v1.ComplianceState>: Compliant
  In [It] at: /go/src/github.com/stolostron/governance-policy-framework/test/integration/policy_gatekeeper_operator_downstream_test.go:148
```

**Comment by JustinKuli** [comment]
/retest

Ah, flakiness

```
GRC: [P1][Sev1][policy-grc] Test the Policy Generator in an App subscription [BVT]
/go/src/github.com/stolostron/governance-policy-framework/test/integration/policy_generator_test.go:23
  [It] Sets up the application subscription
  /go/src/github.com/stolostron/governance-policy-framework/test/integration/policy_generator_test.go:42
  Begin Captured GinkgoWriter Output >>
    STEP: Verifying that the subscription-admin ClusterRoleBinding exists 09/28/22 21:57:29.24
    STEP: Cleaning up the changes made to the cluster in the test 09/28/22 21:57:29.241
    'grc-e2e-policy-generator' namespace still exists: &Namespace{ObjectMeta:{      0 0001-01-01 00:00:00 +0000 UTC <nil> <nil> map[] map[] [] []  []},Spec:NamespaceSpec{Finalizers:[],},Status:NamespaceStatus{Phase:,Conditions:[]NamespaceCondition{},},}
  << End Captured GinkgoWriter Output
  Expected error to be 'already exists': Post "https://api.policy-grc-cp-prow-hj724.dev08.red-chesterfield.com:6443/apis/rbac.authorization.k8s.io/v1/clusterrolebindings": read tcp 10.128.245.63:32960->18.213.57.243:6443: read: connection reset by peer
  Expected
      <bool>: false
  to be true
  In [It] at: /go/src/github.com/stolostron/governance-policy-framework/test/integration/policy_generator_test.go:62
```

**Comment by dhaiducek** [comment]
🙁 I do think we can probably get rid of that Namespace debug line. I added it thinking the namespace was a cause once upon a time, but I think it's unnecessary noise at this point:
https://github.com/stolostron/governance-policy-framework/blob/0f6d7e666a3296dcd36fa9a8bb5be9b43f8fd69b/test/integration/policy_generator_acm_hardening_test.go#L37-L39

**Comment by JustinKuli** [comment]
Fingers crossed that we get the same kind of failure with the new debug stuff...

**Comment by ChunxiAlexLuo** [comment]
> Add big debug for compliance event confusion

Passed this time😂

**Comment by JustinKuli** [comment]
/test test-e2e-grc-framework

Try to force the failure again

**Comment by JustinKuli** [comment]
/retest

```
Claim provision timed out - exiting
...
INFO[2022-09-30T19:02:48Z] Step test-e2e-grc-framework-ocm-e2e-clusterpool-checkout failed after 1h1m50s. 
```

**Comment by JustinKuli** [comment]
/retest

This PR is cursed 😆 

```
Claim provision timed out - exiting
```

**Comment by JustinKuli** [comment]
https://prow.ci.openshift.org/view/gs/origin-ci-test/pr-logs/pull/stolostron_governance-policy-framework/437/pull-ci-stolostron-governance-policy-framework-main-test-e2e-grc-framework/1576926955779919872 had the right failure, but the command to get the logs didn't capture all the information I was looking for... so we'll try again, although from a Slack conversation it sounds like we have a possible fix to work on, based on the understanding of the issue that we already have.

File: test/common/policy_utils.go
Comment (gparvin) [reply test/common/policy_utils.go:78]: The ACM hardening policy set tests with Placement.  That may be it here -- other placement tests are in the propagator.

**Review by dhaiducek — APPROVED** [review]
LGTM! A sign of a good refactor when the amount of lines added is half of those removed! It ought to make future additions/updates much easier too. Nice work!

**Comment by mprahl** [comment]
> /unhold

Please remember to cherry-pick this on 2.6 and 2.5.  :smile:

**Comment by JustinKuli** [comment]
```
Claim provision timed out - exiting
```

/retest 
/cherrypick release-2.5
/cherrypick release-2.6

**Comment by mprahl** [comment]
> ```
> Claim provision timed out - exiting
> ```
> 
> /retest /cherrypick release-2.5 /cherrypick release-2.6

You won't be able to cleanly cherrypick on 2.5 because you are making modifications to tests that don't exist in 2.5, but it should be pretty simple to fix the conflicts.

**Comment by JustinKuli** [comment]
```
Claim provision timed out - exiting
```

/retest

**Comment by JustinKuli** [comment]
```
Claim provision timed out - exiting
```
Still?

/retest

**Comment by JustinKuli** [comment]
```
[2/2] STEP 18/24: RUN ./build/download-clis.sh
which: no oc in (/go/bin:/usr/local/go/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin)
Installing oc and kubectl clis...
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed

  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
100    52  100    52    0     0     22      0  0:00:02  0:00:02 --:--:--    22
100    52  100    52    0     0     22      0  0:00:02  0:00:02 --:--:--    22
gzip: stdin: not in gzip format
tar: Child returned status 1
tar: Error is not recoverable: exiting now
INFO[2022-10-14T16:47:58Z] Ran for 16m12s                               
ERRO[2022-10-14T16:47:58Z] Some steps failed:                           
ERRO[2022-10-14T16:47:58Z] 
  * could not run steps: step grc-policy-framework-tests failed: the build grc-policy-framework-tests failed after 9m35s with reason DockerBuildFailed: Dockerfile build strategy has failed. 
```

Guessing the download of the CLIs got an error. This PR really is cursed...

/retest

**Comment by JustinKuli** [comment]
/cherrypick release-2.6

This failed before, but I think it will work this time because there is one newer commit in the relase-2.6 branch that it needed

**Pull request merged**
